File: data_structures/binary_search_tree/matlab/test_binary_search_tree.m
% test_binary_search_tree.m

t = binary_search_tree('new');
assert(binary_search_tree(t, 'is_empty'));

for v = [5, 3, 7, 1, 4]
    t = binary_search_tree(t, 'insert', v);
end
assert(binary_search_tree(t, 'search', 3));
assert(binary_search_tree(t, 'search', 7));
assert(~binary_search_tree(t, 'search', 6));
assert(binary_search_tree(t, 'size') == 5);

io = binary_search_tree(t, 'inorder');
assert(isequal(io, [1, 3, 4, 5, 7]));

t2 = binary_search_tree('new');
for v = [5, 3, 7]
    t2 = binary_search_tree(t2, 'insert', v);
end
assert(binary_search_tree(t2, 'min') == 3);
assert(binary_search_tree(t2, 'max') == 7);

[t2, ok] = binary_search_tree(t2, 'remove', 3);
assert(ok);
assert(isequal(binary_search_tree(t2, 'inorder'), [5, 7]));
[t2, ok] = binary_search_tree(t2, 'remove', 99);
assert(~ok);

t3 = binary_search_tree('new');
t3 = binary_search_tree(t3, 'insert', 5);
t3 = binary_search_tree(t3, 'insert', 5);
assert(binary_search_tree(t3, 'size') == 1);

t_empty = binary_search_tree('new');
try
    binary_search_tree(t_empty, 'min');
    error('expected error not thrown');
catch e
    assert(strcmp(e.identifier, 'binary_search_tree:underflow'));
end

disp('All tests passed.');

File: data_structures/binary_search_tree/python/test_binary_search_tree.py
import pytest
from binary_search_tree import BinarySearchTree


def test_new_is_empty():
    t = BinarySearchTree()
    assert t.is_empty()
    assert t.size() == 0


def test_insert_and_search():
    t = BinarySearchTree()
    for v in [5, 3, 7, 1, 4]:
        t.insert(v)
    assert t.search(3)
    assert t.search(7)
    assert not t.search(6)
    assert t.size() == 5


def test_inorder_sorted():
    t = BinarySearchTree()
    for v in [5, 3, 7, 1, 4]:
        t.insert(v)
    assert t.inorder() == [1, 3, 4, 5, 7]


def test_preorder():
    t = BinarySearchTree()
    for v in [5, 3, 7]:
        t.insert(v)
    assert t.preorder() == [5, 3, 7]


def test_postorder():
    t = BinarySearchTree()
    for v in [5, 3, 7]:
        t.insert(v)
    assert t.postorder() == [3, 7, 5]


def test_min_max():
    t = BinarySearchTree()
    for v in [5, 3, 7]:
        t.insert(v)
    assert t.min() == 3
    assert t.max() == 7


def test_remove():
    t = BinarySearchTree()
    for v in [5, 3, 7]:
        t.insert(v)
    assert t.remove(3)
    assert t.inorder() == [5, 7]
    assert t.remove(5)
    assert t.inorder() == [7]
    assert not t.remove(99)


def test_duplicate_ignored():
    t = BinarySearchTree()
    t.insert(5)
    t.insert(5)
    assert t.size() == 1


def test_min_empty_raises():
    with pytest.raises(IndexError):
        BinarySearchTree().min()
